### PR TITLE
Add AMDGPU execution provider support

### PR DESCRIPTION
**Pull request by AMDmoore**
﻿## Summary

- Register a new `DeviceType::AMDGPU` in OGA so models can target the **AMD GPU umbrella execution provider** (shipped as `amdgpu-ep.dll` / `libamdgpu-ep.so`). The umbrella EP forwards to its backend (currently the HIP/MorphiZen backend), selected at runtime via the `"profile"` provider option. The integration mirrors the existing RyzenAI path: an `AMDGPU` namespace + `Interface` (allocator, dispatch, session-options setup), a dispatch-table entry in `session_options.cpp`, and a `CreateMemoryInfo_V2` code path in `EnsureDeviceOrtInit` (because the plugin EP's `OrtMemoryInfo` is registered via the V2 API with AMD GPU vendor id 0x1002, not the legacy `OrtMemoryInfo::Create`). Both `past_present_share_buffer=true` and `=false` work; KV cache routing through the AMD GPU allocator is device-agnostic in `DefaultKeyValueCache`.

- Add `DeviceType::AMDGPU` to the `devices_supporting_continuous_decoding` whitelist in `Generator::AppendTokens`. Without it, multiple `AppendTokens` calls without an intervening `RewindToLength(0)` — as used by teacher-forcing PPL evaluation and multi-turn chat — throw "Continuous decoding is not supported on the selected device type". The AMD GPU EP's KV cache lives in UMA-pinned memory (`hipHostMallocMapped`, host-readable) and supports the same shared-buffer KV cache semantics as CUDA / WebGPU / RyzenAI, so it qualifies for the same whitelist entry.

- Allocate AMD GPU model inputs (`input_ids`, `position_ids`, `attention_mask`, etc.) on device memory via the EP's GPU allocator instead of CPU, so ORT can alias the buffers directly (zero H2D/D2H copies for inputs and logits).

## Configuration

```json
"provider_options": [{ "AMDGPU": { "profile": "llm" } }]
```

or

```python
config.append_provider("AMDGPU")
```

## Notes

- New source directory `src/amdgpu/` ports from `src/ryzenai/` with the EP name and vendor filter swapped; behavioural differences are documented inline.
- The EP is registered as `"amdgpu"` and loaded from `amdgpu-ep.dll` (`libamdgpu-ep.so`); the path can be overridden with the `AMDGPU_EP_PATH` environment variable. `SetupProvider` mirrors provider options into `ep.<name>.*` session-config entries since the umbrella reads them from there.
- `EnsureDeviceOrtInit` takes the V2 OrtMemoryInfo path for AMDGPU because the legacy `OrtMemoryInfo::Create` only knows ORT-internal device names ("Cuda", "DML", ...) and would fail with "Specified device is not supported. Try CreateMemoryInfo_V2." for any plugin EP. The device-init dummy session reuses the AMDGPU provider options from `genai_config` so the umbrella can select its backend; no profile is hardcoded.
- The AMD GPU EP itself (the DLL providing the kernels) is shipped separately and is not part of this PR.

### CODE REVIEW

**Comment by AMDmoore** [comment]
> @AMDmoore please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="AMD"

File: src/smartptrs.h [hunk 1]
```diff
@@ -95,6 +95,7 @@ enum struct DeviceType {
   OpenVINO,
   NvTensorRtRtx,
   RyzenAI,
+  MorphiZenEP,
   MAX
```
[the remark this answers is not in the document]
Comment (AMDmoore) [reply src/smartptrs.h:99]: I checked all the references to DeviceType::MorphiZenEP, they are all correct.

File: src/amdgpu/interface.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Generators {
+
+// Memory allocated through the MorphiZen EP interface is treated as both
+// host- and device-accessible because the underlying AMD APU iGPU shares
+// physical memory with the host. Mirrors the RyzenAIInterface pattern; the
+// difference is that this provider targets OrtHardwareDeviceType_GPU + AMD
+// GPU vendor id (0x1002) instead of the NPU.
+struct MorphiZenEPInterface : DeviceInterface {
+  using ProviderOptions = std::vector<std::pair<std::string, std::string>>;
```
[the remark this answers is not in the document]
Comment (AMDmoore) [reply src/amdgpu/interface.h:21]: The missing headers are added.

File: src/amdgpu/interface.cpp [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+  Interface() {
+    ep_path_ = ep_filename_;
+    // If the EP DLL is already loaded by the host, there is nothing to do.
+#if defined(_WIN32)
+    if (GetModuleHandleA(ep_filename_))
+      return;
+#else
+    if (dlopen(ep_filename_, RTLD_NOLOAD | RTLD_NOW))
+      return;
+#endif
```
[the remark this answers is not in the document]
Comment (AMDmoore) [reply src/amdgpu/interface.cpp:87]: Changes were done as requests.

File: src/amdgpu/interface.cpp [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+  void SetupProvider(OrtSessionOptions& session_options, const ProviderOptions& provider_options) override {
+    std::vector<const OrtEpDevice*> supported_devices;
+
+    {
+      const OrtEpDevice* const* devices = nullptr;
+      size_t ndevices = 0;
+
+      Ort::ThrowOnError(Ort::api->GetEpDevices(&GetOrtEnv(), &devices, &ndevices));
+
+      for (const auto& device : std::span{devices, ndevices}) {
```
[the remark this answers is not in the document]
Comment (AMDmoore) [reply src/amdgpu/interface.cpp:150]: It's owned by ORT and will not be released by caller.

File: src/amdgpu/interface.cpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+  }
+
+  void SetupProvider(OrtSessionOptions& session_options, const ProviderOptions& provider_options) override {
+    std::vector<const OrtEpDevice*> supported_devices;
+
+    {
+      const OrtEpDevice* const* devices = nullptr;
+      size_t ndevices = 0;
+
+      Ort::ThrowOnError(Ort::api->GetEpDevices(&GetOrtEnv(), &devices, &ndevices));
```
Comment (kunal-vaishnavi) [inline src/amdgpu/interface.cpp:148]: Can we use the helper APIs from `onnxruntime_inline.h` instead of making direct `Ort::api` calls?

https://github.com/microsoft/onnxruntime-genai/blob/6be530072553f90e1bb438ae21651f1740a080fe/src/models/onnxruntime_inline.h#L138-L140

File: src/amdgpu/interface.cpp [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    // with "ep.<name>.", so mirror them there in addition to passing them through
+    // the plugin EP V2 metadata below.
+    for (auto& option : provider_options) {
+      ep_keys.emplace_back(option.first.c_str());
+      ep_values.emplace_back(option.second.c_str());
+      config_keys.emplace_back(std::string{"ep."} + ep_name_ + "." + option.first);
+    }
+
+    for (size_t i = 0; i < config_keys.size(); ++i) {
+      Ort::ThrowOnError(Ort::api->AddSessionConfigEntry(
```
Comment (kunal-vaishnavi) [inline src/amdgpu/interface.cpp:176]: Same comment as above for the other direct `Ort::api` calls in this PR.

https://github.com/microsoft/onnxruntime-genai/blob/6be530072553f90e1bb438ae21651f1740a080fe/src/models/onnxruntime_inline.h#L678-L680

File: src/models/model.cpp [hunk 7]
```diff
@@ -402,63 +400,65 @@ void EnsureDeviceOrtInit(DeviceInterface& device, const Config& config) {
   // This ensures memory allocated on-device for model inputs/outputs is valid for the lifetime of GenAI.
 
   // Names for the device types used by 'SetProviderSessionOptions'
-  static const char* device_type_names[] = {"CPU (Not used, see above)", "cuda", "DML", "WebGPU", "QNN", "OpenVINO (Not used, see above)", "NvTensorRtRtx", "RyzenAI"};
+  static const char* device_type_names[] = {"CPU (Not used, see above)", "cuda", "DML", "WebGPU", "QNN", "OpenVINO (Not used, see above)", "NvTensorRtRtx", "RyzenAI", "AMDGPU"};
```
Comment (kunal-vaishnavi) [inline src/models/model.cpp:403]: Same comment as [here](https://github.com/microsoft/onnxruntime-genai/pull/2105#discussion_r3496154320) regarding simplification. We could create new APIs such as `GetProviderName()` and `GetMemoryInfoName()` on `DeviceInterface` instead of using parallel string arrays.

File: src/models/model.cpp [hunk 8]
```diff
@@ -402,63 +400,65 @@ void EnsureDeviceOrtInit(DeviceInterface& device, const Config& config) {
...
   allocator.session_ = OrtSession::Create(GetOrtEnv(), g_trivial_model, sizeof(g_trivial_model), session_options.get());
 
   // Names for the device memory types used by 'OrtMemoryInfo::Create'
-  static const char* device_memory_type_names[] = {"CPU (Not used, see above)", "Cuda", "DML", "WebGPU_Buf", "QnnHtpShared", "OpenVINO (Not used, see above)", "Cuda", "Cpu"};
+  static const char* device_memory_type_names[] = {"CPU (Not used, see above)", "Cuda", "DML", "WebGPU_Buf", "QnnHtpShared", "OpenVINO (Not used, see above)", "Cuda", "Cpu", "AMDGPU"};
   static_assert(std::size(device_memory_type_names) == static_cast<size_t>(DeviceType::MAX));
 
   // Get the allocator from the OrtSession for the DeviceType (it's called 'AllocatorCreate' but it's really 'AllocatorGet')
   auto name = device_memory_type_names[static_cast<int>(type)];
   try {
```
Comment (kunal-vaishnavi) [inline src/models/model.cpp:439]: I see several conditionals are getting introduced on a per-EP basis for getting an `OrtMemoryInfo` object. Can we move this per-EP work to the device interface and make it more generic here?

**Comment by kunal-vaishnavi** [comment]
Closing this PR in favor of https://github.com/microsoft/onnxruntime-genai/pull/2165
